### PR TITLE
Enable peer cert trust domain verification with SPIFFE

### PR DESCRIPTION
**Pull request by myidpt**
Please provide a description for what this PR is for.
This PR offers a function to choose the correct cert pool to verify the certificate of a peer corresponding to the peer's trust domain.

This is a follow up work of https://github.com/istio/istio/pull/24669.

Tracking bug: https://github.com/istio/istio/pull/24836

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pkg/spiffe/spiffe_test.go [hunk 1]
```diff
@@ -91,6 +94,190 @@ var (
 		}
 	]
 }`
+
+	validRootCert = `-----BEGIN CERTIFICATE-----
```
Comment (howardjohn) [inline pkg/spiffe/spiffe_test.go:98]: please add a comment of the command to generate this

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -899,35 +903,45 @@ func (s *Server) getCertKeyPaths(tlsOptions TLSOptions) (string, string) {
...
-	cp.AppendCertsFromPEM(rootCertBytes)
 
-	if features.SpiffeBundlePaths != "" {
-		certMap, err := spiffe.RetrieveSpiffeBundleRootCertsFromStringInput(features.SpiffeBundlePaths, []*x509.Certificate{})
+	if len(rootCertBytes) != 0 {
+		block, _ := pem.Decode(rootCertBytes)
+		if block == nil {
+			return fmt.Errorf("failed to decode root cert PEM")
+		}
+		rootCert, err := x509.ParseCertificate(block.Bytes)
```
Comment (JimmyCYJ) [inline pilot/pkg/bootstrap/server.go:928]: Are we guaranteed that there is only one cert in rootCertBytes? Is it possible that there will be a cert pool of root certs?

File: pilot/pkg/bootstrap/server.go [hunk 3]
```diff
@@ -899,35 +903,45 @@ func (s *Server) getCertKeyPaths(tlsOptions TLSOptions) (string, string) {
...
+		certMap, err := spiffe.RetrieveSpiffeBundleRootCertsFromStringInput(
+			features.SpiffeBundleEndpoints, []*x509.Certificate{})
+		if err != nil {
+			return err
 		}
+		s.peerCertVerifier.AddMappings(certMap)
 	}
 
-	return cp, nil
+	return nil
```
Comment (JimmyCYJ) [inline pilot/pkg/bootstrap/server.go:944]: What if features.SpiffeBundleEndpoints is empty and  rootCertBytes is empty? Shall we return error or nil?

File: pkg/spiffe/spiffe.go [hunk 4]
```diff
@@ -221,3 +230,80 @@ func RetrieveSpiffeBundleRootCerts(config map[string]string, extraTrustedCerts [
...
+// It verifies the peer certificate using the root certificates associated with its trust domain.
+func (v *PeerCertVerifier) VerifyPeerCert(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+	if len(rawCerts) == 0 {
+		// Peer doesn't present a certificate. Just skip. Other authn methods may be used.
+		return nil
+	}
+	var peerCert *x509.Certificate
+	intCertPool := x509.NewCertPool()
+	for id, rawCert := range rawCerts {
+		cert, err := x509.ParseCertificate(rawCert)
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe.go:282]: Shall we use ParseCertificates https://golang.org/pkg/crypto/x509/#ParseCertificates?

File: pilot/pkg/bootstrap/server.go (hunk 2 above)
Comment (myidpt) [reply pilot/pkg/bootstrap/server.go:928]: Good question. I think that's possible. I'll create a followup PR to fix this.

File: pilot/pkg/bootstrap/server.go (hunk 3 above)
Comment (myidpt) [reply pilot/pkg/bootstrap/server.go:944]: No, in that case peerCertVerifier will be nil and we won't create a secure GRPC service.

File: pkg/spiffe/spiffe.go (hunk 4 above)
Comment (myidpt) [reply pkg/spiffe/spiffe.go:282]: This is a single cert. I think using ParseCertificate should be fine.

**Pull request merged**